### PR TITLE
Refactor matmul code with mul!

### PR DESCRIPTION
**Pull request by blegat**
The new implementation catches `mul!` calls for which the eltype of the result is a JuMP affine or quadratic expressions (resolving https://github.com/JuliaOpt/JuMP.jl/issues/1990).
This allows the generic codes using `mul!` to benefit from our specialized multiplication methods.
For SparseArrays, the situation is less ideal as some calls are redirected to `mul!` and some have custom implementation that do not work for JuMP as they start by promoting the arrays to the same eltype...
This PR redirects some of them so that all tests pass but more are needed (more tests and more redirections) but this is left for future work. I don't think there is any regression as we already had this issue: https://github.com/JuliaOpt/JuMP.jl/issues/1276

Closes https://github.com/JuliaOpt/JuMP.jl/issues/1990
Closes https://github.com/JuliaOpt/JuMP.jl/issues/1716

### CODE REVIEW

File: src/operators.jl [hunk 1]
```diff
@@ -328,119 +328,44 @@ end
...
     for I in eachindex(A)
-        A[I] = zero(T)
+        A[I] = zero(eltype(A))
     end
     return A
 end
 
 ###############################################################################
-# `_A_mul_B!(ret, A, B)` adds the result of `A*B` into the buffer `ret`. We roll our own
+# `_mul!(ret, A, B)` adds the result of `A*B` into the buffer `ret`. We roll our own
```
Comment (mlubin) [inline src/operators.jl]: s/adds/sets/. The difference is `ret = A*B` vs `ret += A*B`.

File: src/operators.jl [hunk 2]
```diff
@@ -467,78 +392,188 @@ end
...
+    A = parent(adjA)
+    size(A, 2) == size(ret, 1) || throw(DimensionMismatch())
+    size(A, 1) == size(B, 1) || throw(DimensionMismatch())
     size(B, 2) == size(ret, 2) || throw(DimensionMismatch())
-    nzv = A.nzval
-    rv = A.rowval
-    # `ret` is already filled with zeros by `_return_arrayt`.
-    for k = 1:size(ret, 2)
-        @inbounds for col = 1:A.n
+    nzv = nonzeros(A)
```
Comment (mlubin) [inline src/operators.jl]: These names don't follow JuMP conventions: `A_nonzeros`, `A_rowvals` etc

File: src/operators.jl [hunk 3]
```diff
@@ -467,78 +392,188 @@ end
 > OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 > WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 =#
-function _At_mul_B!(ret::StridedVecOrMat{<:GenericAffOrQuadExpr}, A::SparseMatrixCSC, B::StridedVecOrMat)
-    A.n == size(ret, 1) || throw(DimensionMismatch())
-    A.m == size(B, 1) || throw(DimensionMismatch())
+
+const TrAdj{T<:Union{GenericAffOrQuadExpr, Real}, MT} = Union{Transpose{T, MT}, Adjoint{T, MT}}
```
Comment (mlubin) [inline src/operators.jl]: TrAdj is not a meaningful name (is it transposed adjoint?). What about `TransposeOrAdjoint`?

File: src/operators.jl [hunk 4]
```diff
@@ -467,78 +392,188 @@ end
...
+    T = _A_mul_B_eltype(eltype(A), eltype(B))
+    ret = Array{T}(undef, _A_mul_B_ret_dims(A, B))
+    return mul!(ret, A, B)
 end
 
-function Base.:*(A::Union{Matrix, SparseMatrixCSC},
-                 B::Union{Matrix{<:_JuMPTypes}, Vector{<:_JuMPTypes}, SparseMatrixCSC{<:_JuMPTypes}})
-    return _A_mul_B(A, B)
-end
+# A few a overwritten below but many more need to be redirected to `_mul` in
```
Comment (mlubin) [inline src/operators.jl]: s/a/are/

File: src/operators.jl (hunk 1 above)
Comment (blegat) [reply src/operators.jl]: Good point

File: src/operators.jl (hunk 3 above)
Comment (chriscoey) [reply src/operators.jl]: The name is pretty tradj

**Pull request merged**